distributed/protocol/cudf.py

```diff
@@ -0,0 +1,74 @@
+import cudf
+from .serialize import dask_serialize, dask_deserialize
+from .numba import serialize_numba_ndarray, deserialize_numba_ndarray
+
+
+# TODO:
+# 1. Just use positions
+#    a. Fixes duplicate columns
+#    b. Fixes non-msgpack-serializable names
+# 2. cudf.Series
+# 3. Serialize the index
+
+
+@dask_serialize.register(cudf.DataFrame)
+def serialize_cudf_dataframe(x):
+    sub_headers = []
+    arrays = []
+    null_masks = []
+    null_headers = []
+    null_counts = {}
+
+    for label, col in x.iteritems():
+        header, [frame] = serialize_numba_ndarray(col.data.mem)
+        header['name'] = label
+        sub_headers.append(header)
+        arrays.append(frame)
+        if col.null_count:
+            header, [frame] = serialize_numba_ndarray(col.nullmask.mem)
+            header['name'] = label
+            null_headers.append(header)
+            null_masks.append(frame)
+            null_counts[label] = col.null_count
+
+    arrays.extend(null_masks)
+
+    header = {
+        'is_cuda': len(arrays),
+        'subheaders': sub_headers,
+        # TODO: the header must be msgpack (de)serializable.
+        # See if we can avoid names, and just use integer positions.
```

Comment:
Why avoid names?  For fear that they won't be msgpack-serializable?

Reply:
That, and duplicate columns.

@quasiben sorry for the mess, https://github.com/TomAugspurger/distributed/tree/ucx%2Bhacky-serialization has a largeish update to serialization. I thought I had merged it into my `ucx` branch, but apparently not. I think the last 4 commits are relevant, but it's a bit of a mess. https://github.com/TomAugspurger/distributed/commits/ucx%2Bhacky-serialization